gui/widgets/groupedlist.cpp

```diff
@@ -0,0 +1,427 @@
+			checkBounds();
+			markAsDirty();
+
+			// Scrollbar actions cause list focus (which triggers a redraw)
+			// NOTE: ListWidget's boss is always GUI::Dialog
+			((GUI::Dialog *)_boss)->setFocusWidget(this);
+		}
+		break;
+	default:
+		break;
+	}
+}
+
+void GroupedListWidget::reflowLayout() {
+	Widget::reflowLayout();
+
+	_leftPadding = g_gui.xmlEval()->getVar("Globals.ListWidget.Padding.Left", 0);
+	_rightPadding = g_gui.xmlEval()->getVar("Globals.ListWidget.Padding.Right", 0);
+	_topPadding = g_gui.xmlEval()->getVar("Globals.ListWidget.Padding.Top", 0);
+	_bottomPadding = g_gui.xmlEval()->getVar("Globals.ListWidget.Padding.Bottom", 0);
+	_hlLeftPadding = g_gui.xmlEval()->getVar("Globals.ListWidget.hlLeftPadding", 0);
+	_hlRightPadding = g_gui.xmlEval()->getVar("Globals.ListWidget.hlRightPadding", 0);
+
+	_scrollBarWidth = g_gui.xmlEval()->getVar("Globals.Scrollbar.Width", 0);
+
+	// HACK: Once we take padding into account, there are times where
+	// integer rounding leaves a big chunk of white space in the bottom
+	// of the list.
+	// We do a rough rounding on the decimal places of Entries Per Page,
+	// to add another entry even if it goes a tad over the padding.
+	frac_t entriesPerPage = intToFrac(_h - _topPadding - _bottomPadding) / kLineHeight;
+
+	// Our threshold before we add another entry is 0.9375 (0xF000 with FRAC_BITS being 16).
+	const frac_t threshold = intToFrac(15) / 16;
+
+	if ((frac_t)(entriesPerPage & FRAC_LO_MASK) >= threshold)
+		entriesPerPage += FRAC_ONE;
+
+	_entriesPerPage = fracToInt(entriesPerPage);
+	assert(_entriesPerPage > 0);
+
+	if (_scrollBar) {
+		_scrollBar->resize(_w - _scrollBarWidth, 0, _scrollBarWidth, _h, false);
+		scrollBarRecalc();
+		scrollToCurrent();
+	}
+}
+
+void GroupedListWidget::toggleGroup(int groupID) {
+	// Shrink group if it is expanded
+	if (_groupExpanded[groupID]) {
+		_groupExpanded[groupID] = false;
+		sortGroups();
+	} else {
+	// Expand group if it is shrunk
+		_groupExpanded[groupID] = true;
+		sortGroups();
+	}
+}
+
+void GroupedListWidget::drawWidget() {
+	int i, pos, len = _list.size();
+	Common::U32String buffer;
+
+	// Draw a thin frame around the list.
+	g_gui.theme()->drawWidgetBackground(Common::Rect(_x, _y, _x + _w, _y + _h),
+	                                    ThemeEngine::kWidgetBackgroundBorder);
+
+	// Draw the list items
+	for (i = 0, pos = _currentPos; i < _entriesPerPage && pos < len; i++, pos++) {
+		const int y = _y + _topPadding + kLineHeight * i;
+		const int fontHeight = kLineHeight;
+		ThemeEngine::TextInversionState inverted = ThemeEngine::kTextInversionNone;
+
+		// Draw the selected item inverted, on a highlighted background.
+		if (_selectedItem == pos)
+			inverted = _inversion;
+
+		Common::Rect r(getEditRect());
+		int pad = _leftPadding;
+		int rtlPad = (_x + r.left + _leftPadding) - (_x + _hlLeftPadding);
+
+		if (isGroupHeader(_listIndex[pos])) {
+			int groupID = indexToGroupID(_listIndex[pos]);
+			r.left += fontHeight - 2 + _leftPadding;
```

Comment:
Why 2?

Reply:
This came from list.cpp originally where it looks like instead of getting font height directly, it was set to kLineHeight and always used in calculation as (fontHeight - 2), but the kLineHeight itself is a macro defined as g_gui.getFontHeight() + 2. I have corrected this in both groupedlist.cpp and list.cpp in commit 3b01a25.